vicky/src/lib/database/entities/task.rs

```diff
@@ -0,0 +1,415 @@
+
+                    Task {
+                        id: t.id,
+                        display_name: t.display_name.clone(),
+                        status: t.status.as_str().into(),
+                        locks: real_locks,
+                        features: t.features.split("||").map(String::from).collect(),
+                        flake_ref: FlakeRef {
+                            flake: t.flake_ref_uri.clone(),
+                            args: t.flake_ref_args.split("||").map(String::from).collect(),
+                        },
+                    }
+                })
+                .collect();
+
+            Ok(real_tasks)
+        }
+
+        fn get_task(&mut self, tid: Uuid) -> Result<Option<Task>, VickyError> {
+            // so evil >:O
+            use self::locks::dsl::*;
+            use self::tasks::dsl::*;
+
+            let db_task = tasks.filter(self::tasks::id.eq(tid)).first::<DbTask>(self);
+            let db_task = match db_task {
+                Err(diesel::result::Error::NotFound) => return Ok(None),
+                _ => db_task?,
+            };
+            let db_locks: Vec<DbLock> = locks.filter(task_id.eq(task_id)).load::<DbLock>(self)?;
+
+            let task = Task {
+                id: db_task.id,
+                display_name: db_task.display_name.clone(),
+                locks: db_locks.into_iter().map(|l| l.into()).collect(),
+                features: db_task
+                    .features
+                    .split("||")
+                    .map(|s| s.to_string())
+                    .collect(),
+                flake_ref: FlakeRef {
+                    flake: db_task.flake_ref_uri.clone(),
+                    args: db_task
+                        .features
+                        .split("||")
+                        .map(|s| s.to_string())
+                        .collect(),
+                },
+                status: db_task.status.as_str().into(),
+            };
+
+            Ok(Some(task))
+        }
+
+        fn put_task(&mut self, task: &Task) -> Result<(), VickyError> {
+            // even more evil >;(
+            use self::locks::dsl::*;
+            use self::tasks::dsl::*;
+
+            let db_locks: Vec<DbLock> = task
+                .locks
+                .iter()
+                .map(|l| DbLock::from_lock(l, task.id))
+                .collect();
+            let db_task: DbTask = task.into();
+
+            insert_into(tasks).values(db_task).execute(self)?;
+            for db_lock in db_locks {
+                let new_db_lock: NewDbLock = db_lock.into();
+                insert_into(locks).values(new_db_lock).execute(self)?;
+            }
+            Ok(())
+        }
+
+        fn update_task(&mut self, task: &Task) -> Result<(), VickyError> {
+            // even more evil >;(
+            use self::tasks::dsl::*;
+
+            let db_task: DbTask = task.into();
+
+            insert_into(tasks)
```

Comment:
If we do an upsert here, why is there a `put_task` and a `update_task` method with different code?

We might want to fail, if there is an existing task to be consistent or reduce the duplicate here.

Reply:
I'm not even kidding, but before today i was unable to find anything related to updating apart from upsert in diesel. Don't ask me how. Fixed now.